habitat/core/batched_env.py

```diff
@@ -15,6 +15,85 @@
 
+from habitat.utils.geometry_utils import  quaternion_rotate_vector 
+from habitat.tasks.utils import cartesian_to_polar
+
+from habitat_sim.utils.common import  quat_from_magnum
+
+
+class StateSensorConfig:
+    def __init__(self, config_key, shape, obs_key):
+        self.config_key = config_key
+        self.shape = shape
+        self.obs_key = obs_key
+    def get_obs(self, state):
+        pass
+
+def _get_spherical_coordinates(source_position, goal_position, source_rotation):
+    direction_vector = goal_position - source_position
+    source_rotation = quat_from_magnum(source_rotation)
+    direction_vector_agent = quaternion_rotate_vector(
+        # source_rotation.inverse(), direction_vector
+        source_rotation.inverse(), direction_vector
+    )
+    
+    _, phi = cartesian_to_polar(
+            -direction_vector_agent[2], direction_vector_agent[0]
+        )
+    theta = np.arccos(
+        direction_vector_agent[1]
+        / np.linalg.norm(direction_vector_agent)
+    )
+    rho = np.linalg.norm(direction_vector_agent)
+
+
+    if direction_vector.length() == 0.0:
+        # The source is at the same place as the target, theta cannot be calculated
+        theta = 0.0
+    return rho, theta, phi
+
+class RobotStartSensorConfig(StateSensorConfig):
+    def __init__(self):
+        super().__init__("ROBOT_START_RELATIVE", 3, "robot_start_relative")
+    def get_obs(self, state):
+        robot_pos = state.robot_pos
+        robot_rot = state.robot_rotation
+        start_pos = state.robot_start_pos
+        return _get_spherical_coordinates(robot_pos, start_pos, robot_rot)
+
+class RobotTargetSensorConfig(StateSensorConfig):
+    def __init__(self):
+        super().__init__( "ROBOT_TARGET_RELATIVE", 3, "robot_target_relative")
+    def get_obs(self, state):
+        robot_pos = state.robot_pos
+        robot_rot = state.robot_rotation
+        target_pos = state.target_obj_start_pos
+        return _get_spherical_coordinates(robot_pos, target_pos, robot_rot)
+
+class EEStartSensorConfig(StateSensorConfig):
+    def __init__(self):
+        super().__init__( "EE_START_RELATIVE", 3, "ee_start_relative")
+    def get_obs(self, state):
+        ee_pos = state.ee_pos
+        ee_rot = state.ee_rotation
+        start_pos = state.robot_start_pos
+        return _get_spherical_coordinates(ee_pos, start_pos, ee_rot)
+
+class EETargetSensorConfig(StateSensorConfig):
+    def __init__(self):
+        super().__init__( "EE_TARGET_RELATIVE", 3, "ee_target_relative")
+    def get_obs(self, state):
+        ee_pos = state.ee_pos
+        ee_rot = state.ee_rotation
+        target_pos = state.target_obj_start_pos
+        return _get_spherical_coordinates(ee_pos, target_pos, ee_rot)
+
+class JointSensorConfig(StateSensorConfig):
+    def __init__(self):
+        super().__init__("JOINT_SENSOR", 15, "joint_pos")
+    def get_obs(self, state):
+        return state.robot_joint_positions
+
 
```

Comment:
Which sensor will use `state.goal_pos`? How do we convey to the policy where the object needs to be placed? Is that still WIP?

Reply:
Looking back at the doc, we did only mention relative position of "start robot position" and "target position" relative to robot base and to EE. We did not talk about goal position. I can add it very easily but the question is what do we want to use.